### ISSUE TITLE
EGS_AliasTable and EGS_SimpleAliasTable are susceptible to round off errors

### ISSUE THREAD

**Issue by mchamberland** [nrc-cnrc/EGSnrc#24]
I'm not sure if anything can be done to guard against this, but EGS_AliasTable and EGS_SimpleAliasTable are susceptible to round off errors when building the table. [Lines 113-116](https://github.com/nrc-cnrc/EGSnrc/blob/6b1938d3186055017da09bfd37631c8b87c0af32/HEN_HOUSE/egs%2B%2B/egs_alias_table.cpp#L113-L116), and [lines 204-206](https://github.com/nrc-cnrc/EGSnrc/blob/6b1938d3186055017da09bfd37631c8b87c0af32/HEN_HOUSE/egs%2B%2B/egs_alias_table.cpp#L204-L206)  in egs_alias_table.cpp all involve comparing two float variables on which arithmetic operations were applied. 

This might be hard to test, but the following example produces an incorrect table on my system. Generate an EGS_SimpleAliasTable with:

N = 6
f = {1.028976E-5, 0.00019611, 1.028976E-5, 1.028976E-5, 0.00019611, 1.028976E-5}

Uncomment line 221 and a correct table would look like this:
alias table: 0 1 0.142459
alias table: 1 4 1
alias table: 2 1 0.142459
alias table: 3 4 0.142459
alias table: 4 4 1
alias table: 5 4 0.142459

However, on my system I get:
alias table: 0 1 0.142459
alias table: 1 1 1
alias table: 2 1 0.142459
alias table: 3 4 0.142459
alias table: 4 4 1
alias table: 5 5 -0.715083

which is clearly wrong (among other things, the last weight is negative).

Normalizing the weights to 1, i.e. f = {0.023743096, 0.452513809, 0.023743096, 0.023743096, 0.452513809, 0.023743096}, produces the correct table but I'm not sure that this will always be the case.

My compilers are:

```
Marc$ gfortran --version; gcc --version; g++ --version;
GNU Fortran (GCC) 5.2.0
Copyright (C) 2015 Free Software Foundation, Inc.

GNU Fortran comes with NO WARRANTY, to the extent permitted by law.
You may redistribute copies of GNU Fortran
under the terms of the GNU General Public License.
For more information about these matters, see the file named COPYING

Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 7.0.2 (clang-700.1.81)
Target: x86_64-apple-darwin15.2.0
Thread model: posix
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 7.0.2 (clang-700.1.81)
Target: x86_64-apple-darwin15.2.0
Thread model: posix
```

**Comment by mchamberland** [comment]
Not sure if this is relevant or not, but [this page](http://www.keithschwarz.com/darts-dice-coins/) describes a different algorithm to create the alias table, which explicitly addresses numerical instabilities. See the last section ("A Practical Version of Vose's Algorithm"). It also provides an [example implementation](http://www.keithschwarz.com/interesting/code/?dir=alias-method) in Java.

I think that would only be suitable for the EGS_SimpleAliasTable, but for the cases I'm working with, that's sufficient.

**Comment by ftessier** [comment]
Thank you @mchamberland for pointing out this problem; we ought to clean this up indeed! Kudos for the useful links as well.

**Comment by mchamberland** [comment]
Here's a sample C++ implementation I whipped up this morning, based on the Java code above. It requires the use of std::stack. Feel free to use it, if helpful.

```
#include <stack>
void simple_alias_table(int size, float *probabilities) {
	int *alias = new int [size];
	float *prob = new float [size];
	std::stack<int> small_stack;
	std::stack<int> large_stack;
	double sum = 0.;
	for (int i = 0; i < size; i++) {
		sum += probabilities[i];
	}
	double average = 1. / size;

	for (int i = 0; i < size; i++) {
		probabilities[i] /= sum;
		if (probabilities[i] < average) {
			small_stack.push(i);
		}
		else {
			large_stack.push(i);
		}
	}
	
	int less = 0;
	int more = 0;
	while (!small_stack.empty() && !large_stack.empty()) {
		less = small_stack.top();
		small_stack.pop();
		more = large_stack.top();
		large_stack.pop();
		
		prob[less] = probabilities[less] * size;
		alias[less] = more;
		probabilities[more] = probabilities[more] + probabilities[less] - average;
		if (probabilities[more] < average) {
			small_stack.push(more);
		}
		else {
			large_stack.push(more);
		}
	}
	
	while (!large_stack.empty()) {
		more = large_stack.top();
		large_stack.pop();
		prob[more] = 1.;
	}
	
	while (!small_stack.empty()) {
		less = small_stack.top();
		small_stack.pop();
		prob[less] = 1.;
	}
```

**Comment by ftessier** [comment]
Fixed in #557.
